torchdata/dataloader2/reading_service.py

```diff
@@ -145,6 +154,10 @@ def init_datapipe_process(num_workers, worker_id, datapipe):
         torch.utils.data.graph_settings.apply_sharding(datapipe, num_workers, worker_id)
 
     def initialize(self, datapipe: DataPipe) -> DataPipe:
+        try:
+            self._length = len(datapipe)
+        except TypeError:
+            pass
```

Comment:
Two things:
1. This is  taking the `len` before any `apply_sharding` happens. I think that is the right number that we want to return for multiprocessing. Agree?
2. This is only ran once (instead of per iteration) as I don't think users want different length per iteration. Is this a reasonable assumption?